snuba/admin/clickhouse/common.py

```diff
@@ -0,0 +1,66 @@
+from typing import MutableMapping
+
+from snuba import settings
+from snuba.clickhouse.native import ClickhousePool
+from snuba.clusters.cluster import ClickhouseClientSettings, ClickhouseCluster
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.storages.factory import get_storage
+from snuba.utils.serializable_exception import SerializableException
+
+
+class InvalidNodeError(SerializableException):
+    pass
+
+
+class InvalidCustomQuery(SerializableException):
+    pass
+
+
+class InvalidStorageError(SerializableException):
+    pass
+
+
+def is_valid_node(host: str, port: int, cluster: ClickhouseCluster) -> bool:
+    nodes = cluster.get_local_nodes()
+    return any(node.host_name == host and node.port == port for node in nodes)
+
+
+CONNECTIONS: MutableMapping[StorageKey, ClickhousePool] = {}
+
+
+def get_clickhouse_connection(
+    clickhouse_host: str, clickhouse_port: int, storage_name: str
+) -> ClickhousePool:
+    storage_key = None
+    try:
+        storage_key = StorageKey(storage_name)
+    except ValueError:
+        raise InvalidStorageError(
+            f"storage {storage_name} is not a valid storage name",
+            extra_data={"storage_name": storage_name},
+        )
+
+    if storage_key in CONNECTIONS:
+        return CONNECTIONS[storage_key]
```

Comment:
I don't think this will work. 

The issue is that there are two separate use cases, you cannot use the same logic to retrieve a connection for the tracing tool (that needs to talk to a query node through envoy) and for the system tables tool (which needs to access individual nodes and uses the storage key just to fetch the right cluster to validate the host name/port).

Specifically:
Here the system table tool will create a connection to the required host the first time and cache it. Subsequently it will always return the same connection to the same host for the same storage, no matter which host/port the user wants. Which is wrong. If you want to cache connections for the system tables tool, the key must contain at least the host name as well.

I think you want two methods, one to return a query connection to a cluster given a storage key. And another one to take host/port and validates they are part of a storage and then return a connection to those. 
The system tables tool would use the first. The tracing tool would use the second (or both if you want to get a trace from a query ran on a storage node)

Reply:
I split the connection functions into two, one for system queries that cache by hostname, and one for tracing queries that cache by cluster.